TAO/TAO_IDL/util/utl_global.cpp

```diff
@@ -1930,63 +1930,90 @@ struct yy_buffer_state;
+#define BUILTIN_NAME_BUFFER_SIZE 64
+        static char buffer[BUILTIN_NAME_BUFFER_SIZE];
+        static unsigned n = 1;
+        ACE_OS::snprintf (&buffer[0], BUILTIN_NAME_BUFFER_SIZE, "builtin-%u", n++);
+#undef BUILTIN_NAME_BUFFER_SIZE
+        UTL_String utl_string (&buffer[0], true);
+        idl_global->idl_src_file (new UTL_String (&utl_string, true));
+        idl_global->set_filename (new UTL_String (&utl_string, true));
+
+        // Disable Output
+        if (disable_output)
+        {
+          default_streambuf = ACE_DEFAULT_LOG_STREAM->rdbuf ();
+          ACE_DEFAULT_LOG_STREAM->rdbuf (0);
+          ACE_LOG_MSG->clr_flags (ACE_Log_Msg::STDERR);
+        }
+      }
+
+      ~OldState()
+      {
+        idl_global->set_lineno (old_lineno);
+
+        // Restore IDL_Global Context
+        idl_global->set_filename (old_filename);
+        idl_global->idl_src_file()->destroy ();
+        delete idl_global->idl_src_file ();
+        idl_global->idl_src_file (old_idl_src_file);
+        idl_global->reset_flag_seen ();
+
+        // Renable Output
+        if (disable_output)
+        {
+          ACE_DEFAULT_LOG_STREAM->rdbuf (default_streambuf);
+          ACE_LOG_MSG->set_flags (flags);
+        }
+      }
+
+  private:
+      IDL_GlobalData* idl_global;
+      UTL_String *old_filename;
+      long old_lineno;
+      UTL_String *old_idl_src_file;
+      bool disable_output;
+      std::streambuf *default_streambuf;
+      const unsigned long flags;
+  };
+}
+
 void
 IDL_GlobalData::eval (const char *string, bool disable_output)
 {
-  in_eval_ = true;
-
-  // Get IDL_Global Context
-  UTL_String *old_filename = filename ();
-  pd_filename = 0;
-  long old_lineno = lineno ();
-  idl_global->set_lineno (1);
-  UTL_String *old_idl_src_file = idl_src_file ();
+  exit(0);
+  OldState old(this, disable_output);
 
-  // Name this pseudo-file "builtin-N"
-#define BUILTIN_NAME_BUFFER_SIZE 64
-  static char buffer[BUILTIN_NAME_BUFFER_SIZE];
-  static unsigned n = 1;
-  ACE_OS::snprintf (&buffer[0], BUILTIN_NAME_BUFFER_SIZE, "builtin-%u", n++);
-#undef BUILTIN_NAME_BUFFER_SIZE
-  UTL_String utl_string (&buffer[0], true);
-  idl_global->idl_src_file (new UTL_String (&utl_string, true));
-  idl_global->set_filename (new UTL_String (&utl_string, true));
+  in_eval_ = true;
 
   // Set up Flex to read from string
   tao_yy_scan_string (string);
 
-  // Disable Output
-  std::streambuf *default_streambuf = ACE_DEFAULT_LOG_STREAM->rdbuf ();
-  const unsigned long flags = ACE_LOG_MSG->flags ();
-  if (disable_output)
-    {
-      ACE_DEFAULT_LOG_STREAM->rdbuf (0);
-      ACE_LOG_MSG->clr_flags (ACE_Log_Msg::STDERR);
-    }
-
   // emulate DRV_drive()
   FE_yyparse ();
   idl_global->check_primary_keys ();
   AST_check_fwd_decls ();
 
-  // Renable Output
-  if (disable_output)
-    {
-      ACE_DEFAULT_LOG_STREAM->rdbuf (default_streambuf);
-      ACE_LOG_MSG->set_flags (flags);
-    }
-
   // Have Flex Cleanup
   tao_yylex_destroy ();
```

Comment:
Why did you not put this or `in_eval_ = false` in the destructor? Both of these are part of the cleanup.

Reply:
in_eval_ is a public member in IDL_GlobalData. I could pass it as a reference to the helper class so it could be accessed there. it that what you meant?

Reply:
`idl_global->in_eval_ ` should be fine.